### PR TITLE
[SCFToCalyx] Add source location metadata for Cider.

### PR DESCRIPTION
**Pull request by cgyurgyik**
This writes source location information for each leaf node of the control in a format that can be parsed by Cider.

- [Calyx #974](https://github.com/cucapra/calyx/issues/974)
- [Cider documentation](https://docs.calyxir.org/debug/cider.html)

```mlir
// test.mlir
module {
  func @main(%a0 : i32, %a1 : i32, %a2 : i32) -> i32 {
    %0 = arith.addi %a0, %a1 : i32
    %1 = arith.addi %0, %a1 : i32
    %b = arith.cmpi uge, %1, %a2 : i32
    cf.cond_br %b, ^bb1, ^bb2
  ^bb1:
    return %a1 : i32
  ^bb2:
    return %a2 : i32
  }
}
```

->

```mlir
 // ./circt-opt test.mlir --lower-scf-to-calyx='cider-debug-path=metadata.txt' -canonicalize | ./circt-translate --export-calyx
import "primitives/core.futil";
component main<"toplevel"=1>(in0: 32, in1: 32, in2: 32, @clk clk: 1, @reset reset: 1, @go go: 1) -> (out0: 32, @done done: 1) {
  cells {
    std_ge_0 = std_ge(32);
    std_add_1 = std_add(32);
    std_add_0 = std_add(32);
    ret_arg0_reg = std_reg(32);
  }
  wires {
    out0 = ret_arg0_reg.out;
    comb group bb0_2 {
      std_ge_0.left = std_add_1.out;
      std_ge_0.right = in2;
      std_add_1.left = std_add_0.out;
      std_add_0.left = in0;
      std_add_0.right = in1;
      std_add_1.right = in1;
    }
    group ret_assign_0 {
      ret_arg0_reg.in = in1;
      ret_arg0_reg.write_en = 1'd1;
      ret_assign_0[done] = ret_arg0_reg.done;
    }
    group ret_assign_1 {
      ret_arg0_reg.in = in2;
      ret_arg0_reg.write_en = 1'd1;
      ret_assign_1[done] = ret_arg0_reg.done;
    }
  }
  control {
    seq {
      if std_ge_0.out with bb0_2 {
        seq {
          @tag(0) ret_assign_0;
        }
      }
      else {
        seq {
          @tag(1) ret_assign_1;
        }
      }
    }
  }
}
```

```
// metadata.txt
0: loc("test.mlir":10:5)
1: loc("test.mlir":12:5)
```

### CODE REVIEW

**Comment by cgyurgyik** [comment]
CC @EclecticGriffin

**Comment by rachitnigam** [comment]
Can you summarize when the `pos` attributes are generated? Is each basic block going to generate one or can there be more?

**Comment by cgyurgyik** [comment]
Currently, we only generate `pos` attributes after all passes have been run. Unlike LLVM, MLIR requires that a source location is passed in when building a new value. This is demonstrated here:

https://github.com/llvm/circt/blob/38b05d13a6ab718c5c866ae50e9e8ca0b0dc5688/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L243-L248

Since SCF basic blocks may produce 1 or more groups, e.g., a read followed by a write to memory, each may produce 1 or more `pos` tags. These `pos` tags are associated with IR within the basic block, and not the basic block itself. Any discrepancy in the source location is likely a bug in the code.

A follow-up step will probably be to test this on more complex cases.

**Comment by cgyurgyik** [comment]
Friendly bump on this. @rnigam @EclecticGriffin

**Comment by rachitnigam** [comment]
LGTM! Can you make sure that the calyx compiler can parse the output metadata? If that works, merge!

**Comment by cgyurgyik** [comment]
> LGTM! Can you make sure that the calyx compiler can parse the output metadata? If that works, merge!

This should be equivalent to the syntax in https://github.com/cucapra/calyx/pull/996. I haven't tried actually running it through the Calyx compiler but that is the next step, so I'll fix it if there's any minor errors. I just don't want to block any SCF/StaticLogic refactoring.

**Comment by rnigam** [comment]
Hello fellow googlers, I should probably be taken off this list. I dont
think I am the right recipient for this email and this is my personal email.

**Comment by cgyurgyik** [comment]
Sorry, I put the wrong username!
